sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -230,25 +229,43 @@ public void visitPrimitiveTransform(Node node) {
 
   private void replace(final PTransformOverride override) {
     final Collection<Node> matches = new ArrayList<>();
+    final Set<String> freedNames = new HashSet<>();
     transforms.visit(
         new PipelineVisitor.Defaults() {
+          Node clearingNamesFor = null;
+
           @Override
           public CompositeBehavior enterCompositeTransform(Node node) {
+            if (clearingNamesFor != null) {
+              freedNames.add(node.getFullName());
+              return CompositeBehavior.ENTER_TRANSFORM;
+            }
             if (!node.isRootNode() && override.getMatcher().matches(node.toAppliedPTransform())) {
               matches.add(node);
               // This node will be replaced. It should not be visited.
-              return CompositeBehavior.DO_NOT_ENTER_TRANSFORM;
+              clearingNamesFor = node;
+              freedNames.add(node.getFullName());
```

Comment:
Right now there's no way to traverse an arbitrary subgraph of a `TransformHierarchy` or `Pipeline`;

Instead I've changed this to track if the parent node is to-be-freed, starting with any replacement node. I think it's significantly more readable than the nullable variable, but still requires continuing to traverse the graph after we've matched a node to replace. Alternatively I could split this into two traversals. WDYT?